batcher/sweep_batch.go

```diff
@@ -0,0 +1,1148 @@
+		// Sanity check, there should be exactly 1 sweep in this batch.
+		if len(sweeps) != 1 {
+			log.Warnf("Batch %x: non-wallet sweep batched with "+
+				"other sweeps", b.ID[:6])
+		}
+
+		destAddr = sweeps[0].DestAddr
+	} else {
+		var err error
+
+		// Generate a wallet address for the batch transaction's output.
+		destAddr, err = b.wallet.NextAddr(
+			ctx, "", walletrpc.AddressType_TAPROOT_PUBKEY, false,
+		)
+		if err != nil {
+			return fee, err, false
+		}
+	}
+
+	batchPkScript, err := txscript.PayToAddrScript(destAddr)
+	if err != nil {
+		return fee, err, false
+	}
+
+	weightEstimate.AddP2TROutput()
+
+	totalWeight := int64(weightEstimate.Weight())
+
+	fee = b.RbfCache.FeeRate.FeeForWeight(totalWeight)
+
+	// Add the batch transaction output, which excludes the fees paid to
+	// miners.
+	batchTx.AddTxOut(&wire.TxOut{
+		PkScript: batchPkScript,
+		Value:    int64(batchAmt - fee),
+	})
+
+	packet, err := psbt.NewFromUnsignedTx(batchTx)
+	if err != nil {
+		return fee, err, false
+	}
+
+	prevOuts := make(map[wire.OutPoint]*wire.TxOut)
+
+	for i, sweep := range sweeps {
+		txOut := &wire.TxOut{
+			Value:    int64(sweep.Value),
+			PkScript: sweep.Htlc.PkScript,
+		}
+
+		prevOuts[sweep.Outpoint] = txOut
+		packet.Inputs[i].WitnessUtxo = txOut
+	}
+
+	var psbtBuf bytes.Buffer
+	err = packet.Serialize(&psbtBuf)
+	if err != nil {
+		return fee, err, false
+	}
+
+	prevOutputFetcher := txscript.NewMultiPrevOutFetcher(prevOuts)
+
+	for i, sweep := range sweeps {
+		sweep := sweep
+
+		sigHashes := txscript.NewTxSigHashes(
+			packet.UnsignedTx, prevOutputFetcher,
+		)
+
+		sigHash, err := txscript.CalcTaprootSignatureHash(
+			sigHashes, txscript.SigHashDefault, packet.UnsignedTx,
+			i, prevOutputFetcher,
+		)
+		if err != nil {
+			return fee, err, false
+		}
+
+		var (
+			signers       [][]byte
+			muSig2Version input.MuSig2Version
+		)
+
+		// Depending on the MuSig2 version we either pass 32 byte
+		// Schnorr public keys or normal 33 byte public keys.
+		if sweep.ProtocolVersion >= loopdb.ProtocolVersionMuSig2 {
+			muSig2Version = input.MuSig2Version100RC2
+			signers = [][]byte{
+				sweep.HtlcKeys.SenderInternalPubKey[:],
+				sweep.HtlcKeys.ReceiverInternalPubKey[:],
+			}
+		} else {
+			muSig2Version = input.MuSig2Version040
+			signers = [][]byte{
+				sweep.HtlcKeys.SenderInternalPubKey[1:],
+				sweep.HtlcKeys.ReceiverInternalPubKey[1:],
+			}
+		}
+
+		htlcScript, ok := sweep.Htlc.HtlcScript.(*swap.HtlcScriptV3)
+		if !ok {
+			return fee, fmt.Errorf("invalid htlc script version"),
+				false
+		}
+
+		// Now we're creating a local MuSig2 session using the receiver
+		// key's key locator and the htlc's root hash.
+		musig2SessionInfo, err := b.signerClient.MuSig2CreateSession(
+			ctx, muSig2Version,
+			&sweep.HtlcKeys.ClientScriptKeyLocator, signers,
+			lndclient.MuSig2TaprootTweakOpt(
+				htlcScript.RootHash[:], false,
+			),
+		)
+		if err != nil {
+			return fee, err, false
+		}
+
+		if b.muSig2SignSweep == nil {
```

Comment:
We could check only once per runtime, as the method is handed over only once. We need this check to avoid a possible panic if method was not provided.

Reply:
So if we set it at construction time and check there it is not sufficient?

Reply:
the result of that check would then have to be stored & passed down to the publish method in order to not attempt a coop-batch. IMO it's simpler to just check within `publishBatchCoop` and make it error out

Reply:
I'm not quite sure I follow why we'd need to pass down the result of the check? Apart from not properly constructing the batch with the batch kit, in what other case would the signer not be available?